Source/bmalloc/libpas/src/test/EnumerationTests.cpp

```diff
@@ -0,0 +1,315 @@
+ */
+
+#include <map>
+#include <stdlib.h>
+#include <set>
+#include <sys/mman.h>
+
+#include "TestHarness.h"
+#include "pas_root.h"
+#include "bmalloc_heap.h"
+#include "pas_probabilistic_guard_malloc_allocator.h"
+#include "pas_heap.h"
+#include "iso_heap.h"
+#include "iso_heap_config.h"
+#include "pas_heap_ref_kind.h"
+
+
+using namespace std;
+
+namespace {
+
+const bool verbose = false;
+
+struct PageRange {
+    PageRange() = default;
+
+    PageRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(pas_is_aligned(reinterpret_cast<uintptr_t>(base), pas_page_malloc_alignment()));
+        PAS_ASSERT(pas_is_aligned(size, pas_page_malloc_alignment()));
+        PAS_ASSERT(!!base == !!size);
+    }
+
+    bool operator<(PageRange range) const
+    {
+        return base < range.base;
+    }
+
+    void* end() const
+    {
+        return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(base) + size);
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+set<PageRange> pageRanges;
+
+struct RecordedRange {
+    RecordedRange() = default;
+
+    RecordedRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(base);
+    }
+
+    bool operator<(RecordedRange other) const
+    {
+        return base < other.base;
+    }
+
+    void* end() const
+    {
+        return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(base) + size);
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+struct ReaderRange {
+    ReaderRange() = default;
+
+    ReaderRange(void* base, size_t size)
+        : base(base)
+        , size(size)
+    {
+        PAS_ASSERT(base);
+        PAS_ASSERT(size);
+    }
+
+    bool operator<(ReaderRange other) const
+    {
+        if (base != other.base)
+            return base < other.base;
+        return size < other.size;
+    }
+
+    void* base { nullptr };
+    size_t size { 0 };
+};
+
+map<pas_enumerator_record_kind, set<RecordedRange>> recordedRanges;
+map<ReaderRange, void*> readerCache;
+
+void* enumeratorReader(pas_enumerator* enumerator,
+                       void* address,
+                       size_t size,
+                       void* arg)
+{
+    CHECK(!arg);
+    CHECK(size);
+
+    ReaderRange range = ReaderRange(address, size);
+
+    auto readerCacheIter = readerCache.find(range);
+    if (readerCacheIter != readerCache.end())
+        return readerCacheIter->second;
+
+    void* result = pas_enumerator_allocate(enumerator, size);
+
+    void* pageAddress = reinterpret_cast<void*>(
+        pas_round_down_to_power_of_2(
+            reinterpret_cast<uintptr_t>(address),
+            pas_page_malloc_alignment()));
+    size_t pagesSize =
+        pas_round_up_to_power_of_2(
+            reinterpret_cast<uintptr_t>(address) + size,
+            pas_page_malloc_alignment())
+        - reinterpret_cast<uintptr_t>(pageAddress);
```

Comment:
Ditto: These definitions are not being done in standard WebKit style. Does `libpas` have its own style?

Reply:
Yes, it does have some different style choices... but I don't think it would be bad to more closely align with WebKit at this point.